src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -409,6 +402,9 @@ bool BlockchainSQLite::pop_block(const cryptonote::block& block, const service_n
   {
     SQLite::Transaction transaction{db};
 
+    // Add back to the database payments that had been made in this block
+    delete_block_payments(block_height);
+
```

Comment:
I know that this bug fixed the issue, but the change looks *wrong* to me, which makes me suspect it may be glossing over another bug somewhere else.  The problem is, when we add rewards and payments we're doing things in this order:

- calculate any payments based on current accrued rewards, subtract them from accrued amounts.
- add new accrued rewards for the current block.

And so logically, when we *undo* it, we should be undoing in the reverse order and should be doing the exact negatives of the calculations that we entered in the first place:

- subtract accrued rewards for the current block.
- add paid amounts back to accrued rewards.

That's what the version *before* this commit was doing, and so it isn't at all clear to me that moving this line should change anything -- and if it is, it feels like it's merely avoiding hitting a bug somewhere else rather than fixing that other latent bug.

Reply:
> My suspicion is that it's the reduction in value to 0 (which then hits a trigger that deletes the row) that is causing conflict here.

Yes, it's definitely see.  See comments on the trigger.

Reply:
Yeah you nailed it. Although This change fixed the ordering which is why it worked. The ordering is slightly different from what you have listed:

- Calculate payments based on current accrued rewards
- Add accrued rewards for current block to DB
- Subtract payments calculated in step 1 from DB

Probably unimportant at this point, cause i we want to update the triggers also to make sure both work but adding a block changes the database like so: 

10 -> 10.005 -> 0.005

So the current commit works because it reverses this.

Its less of an issue with this current DB structure but previously when we were maintaining the heights getting this ordering wrong caused massive issues.

But yeah overall caused by the trigger not being able to create a new entry if it doesnt already exist. Which is definitely should be able to do.

Reply:
Related: should we switch that order, to add the accrued rewards first and *then* calculate the payments?  So that you if you are paid in block 12345, that includes the payment for 12345 itself?

Reply:
> - Calculate payments based on current accrued rewards
> - Add accrued rewards for current block to DB
> - Subtract payments calculated in step 1 from DB

Are you sure of this order?  I'm looking at `BlockchainSQLite::add_block` and the first thing it does before dealing with rewards is to call `validate_batch_payment` with `save_payment=true`, which calls save_payments() which issues the INSERT into `_paid`.